### PR TITLE
noahdarveau/Rollup changes for treeshaking

### PR DESCRIPTION
**Pull request by noahdarveau-MSFT**
For more information about how to contribute to this repo, visit [this page](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md).

## Description

> Summarize the changes, including the goals and reasons for this change. Be sure to call out any technical or behavior changes that reviewers should be aware of.

> If this Pull Request should close/resolve any issues when merged, use [the special syntax for that](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) here.

### Main changes in the PR:

1. Added Rollup to the repository which will be used to build a treeshakable version of Teams-JS once the treeshaking code changes go in.
2. Changed the pkg.json file to have the `main` field point to the minified cdn bundle and added the `module` field to point to the esm bundle.
3. Updated the Build-Test-Publish pipeline so the CDN package does not included the Rollup-built code.
4. Added a basic tree shaking test app that was used to verify tree shaking is working in teams-js. Note that there are currently no automated tests to check if tree shaking is working and verification must be done manually by examining the bundled package built with the test app.

### Validation performed:

1. All unit and e2e tests continue to pass.

### CODE REVIEW

File: tools/yaml-templates/build-test-publish.yml [hunk 1]
```diff
@@ -118,10 +118,10 @@ steps:
 
   - task: CopyFiles@2
     inputs:
-      Contents: |
-        packages\teams-js\dist\**\?(*.js|*.ts|*.map)
+      sourceFolder: 'packages\teams-js\dist'
```
Comment (jadahiya-MSFT) [inline tools/yaml-templates/build-test-publish.yml:121]: would this include the packages in `umd`?

Comment (noahdarveau-MSFT) [reply tools/yaml-templates/build-test-publish.yml:121]: Yes, it will include both the umd and esm packages. We still need to publish both the umd package to npm to backwards support any apps that are not configured to use es modules

File: packages/teams-js/rollup.config.mjs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import terser from '@rollup/plugin-terser';
+import typescript from '@rollup/plugin-typescript';
+import nodePolyfills from 'rollup-plugin-polyfill-node';
+
+import version from './package.json' assert { type: 'json' };
+
+export default {
+  input: './src/index.ts',
+  output: {
+    dir: 'dist/esm',
```
Comment (jadahiya-MSFT) [inline packages/teams-js/rollup.config.mjs:16]: How is `esm` being used when it comes to publishing?

Comment (jadahiya-MSFT) [reply packages/teams-js/rollup.config.mjs:16]: Your above answer, answers this

Comment (noahdarveau-MSFT) [reply packages/teams-js/rollup.config.mjs:16]: Can you please elaborate a bit more? I'm not sure I fully understand what you're asking

File: tools/yaml-templates/build-test-publish.yml [hunk 3]
```diff
@@ -118,10 +118,10 @@ steps:
...
+      contents: '**\?(*.js|*.ts|*.map)'
       TargetFolder: '$(Build.ArtifactStagingDirectory)\NPMFeed\dist'
-      flattenFolders: true
+      flattenFolders: false
```
Comment (jadahiya-MSFT) [inline tools/yaml-templates/build-test-publish.yml:124]: I assume this is to for backcompat in case a developer doesn't have ESModules configured

File: packages/teams-js/rollup.config.mjs (hunk 2 above)
Comment (jadahiya-MSFT) [reply packages/teams-js/rollup.config.mjs:16]: Oh I was just curious how the `esm` bits were being pulled in and published and your answer about publishing both `esm` and `umd` when publishing to the NPM Feed answered that for me

File: tools/yaml-templates/build-test-publish.yml (hunk 3 above)
Comment (noahdarveau-MSFT) [reply tools/yaml-templates/build-test-publish.yml:124]: Yes, that is one reason, but mostly it is because the rollup-built package keeps all of the files separate, so if we flatten all of the files to one folder level, our published artifact would be a huge mess of every file in teams-js condensed to a single folder level. Keeping the folder structure helps keep the artifact neat.

Comment (jadahiya-MSFT) [reply tools/yaml-templates/build-test-publish.yml:124]: Got it, thanks for the explanation

File: apps/tree-shaking-test-app/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import { barCode } from 'testlibraryfortreeshaking';
+//import { PrintTestMessage } from 'testlibraryfortreeshaking/app';
+import { geoLocation } from '@microsoft/teams-js';
+//import { app } from '@microsoft/teams-js';
```
Comment (jadahiya-MSFT) [inline apps/tree-shaking-test-app/src/index.ts:4]: Do we want to keep these commented out? If so can we have an added comment explaining why

Comment (noahdarveau-MSFT) [reply apps/tree-shaking-test-app/src/index.ts]: Oh, yeah, I forgot about that file. The whole tree-shaking test app isn't actually doing anything. I was just using it for testing, and wanted to check it in for use later. I'll go ahead and remove the comments so it looks nicer.

**Review by jadahiya-MSFT — APPROVED** [review]

File: packages/teams-js/src/private/index.ts [hunk 5]
```diff
@@ -18,7 +18,7 @@ export {
   registerUserSettingsChangeHandler,
   openFilePreview,
 } from './privateAPIs';
-export { conversations } from './conversations';
+export { conversations, OpenConversationRequest, ConversationResponse } from './conversations';
```
Comment (TrevorJoelHarris) [inline packages/teams-js/src/private/index.ts:21]: What is this change needed for?

**Review by TrevorJoelHarris — COMMENTED** [review]
Had one question about a change to the `private\index.ts` file

File: packages/teams-js/src/private/index.ts (hunk 5 above)
Comment (noahdarveau-MSFT) [reply packages/teams-js/src/private/index.ts]: While developing, I was running into a bug where OpenConversationRequest and ConversationResponse was not able to be imported by the teams test app, so I had to export it directly to resolve the issue. I just tried removing the change and it seems to be working ok, so I've reverted the change.

**Review by jadahiya-MSFT — APPROVED** [review]

**Review by TrevorJoelHarris — APPROVED** [review]
